src/home_robot/home_robot/mapping/voxel.py

```diff
@@ -25,88 +53,470 @@ def combine_point_clouds(
+        self.grid_resolution = grid_resolution
+        self._seq = 0
+        self._2d_last_updated = -1
+        self.min_depth = min_depth
+        self.max_depth = max_depth
+
+        # Create an instance memory to associate bounding boxes in space
+        self.instances = InstanceMemory(
+            num_envs=1, du_scale=1, instance_association="bbox_iou"
+        )
+
+        # Create disk for mapping explored areas near the robot - since camera can't always see it
+        self._disk_size = np.ceil(1.0 / self.grid_resolution)
+        self._visited_disk = create_disk(
+            1.0 / self.grid_resolution, (2 * self._disk_size) + 1
+        )
+
+        # Add points with local_radius to the voxel map at (0,0,0) unless we receive lidar points
+        self.add_local_radius_points = add_local_radius_points
+        self.local_radius = local_radius
+
+        if grid_size is not None:
+            self.grid_size = [grid_size[0], grid_size[1]]
+        else:
+            self.grid_size = DEFAULT_GRID_SIZE
+        # Track the center of the grid - (0, 0) in our coordinate system
+        # We then just need to update everything when we want to track obstacles
+        self.grid_origin = np.array(self.grid_size + [0]) // 2
+        # Init variables
+        self.reset()
+
+    def reset(self) -> None:
+        """Clear out the entire voxel map."""
+        self.observations = []
+        # Create map here - just reset *some* variables
+        self.reset_cache()
+
+    def reset_cache(self):
+        """Clear some tracked things"""
+
+        # Stores points in 2d coords where robot has been
+        self._visited = np.zeros(self.grid_size)
+
+        # Store instances detected (all of them for now)
+        self.instances.reset()
+
+        # Store 2d map information
+        # This is computed from our various point clouds
+        self._map2d = None
+
+        # Holds 3d data
+        self._pcd = None
         self.xyz = None
+        self.rgb = None
         self.feats = None
-        self.observations = []
 
-    def add(self, camera_pose: np.ndarray, xyz: np.ndarray, feats: np.ndarray, **info):
-        """Add this to our history of observations. Also update the current running map."""
-        assert xyz.shape[-1] == 3
-        assert feats.shape[-1] == self.feature_dim
-        assert xyz.shape[0] == feats.shape[0]
-        self.observations.append((camera_pose, xyz, feats, info))
-        world_xyz = trimesh.transform_points(xyz, camera_pose)
+    def get_instances(self) -> List[InstanceView]:
+        """Return a list of all viewable instances"""
+        return self.instances.instance_views[0].values()
+
+    def add(
+        self,
+        camera_pose: np.ndarray,
+        xyz: np.ndarray,
+        rgb: np.ndarray,
+        feats: Optional[np.ndarray] = None,
+        depth: Optional[np.ndarray] = None,
+        base_pose: Optional[np.ndarray] = None,
+        obs: Optional[Observations] = None,
```

Comment:
Is obs truly optional here if we need it for the instance mask?

Reply:
technically you dont need instance masks right?